lib/kubeclient/watch_stream.rb

```diff
@@ -1,98 +1,70 @@
 # frozen_string_literal: true
 
 require 'json'
-require 'http'
+
 module Kubeclient
   module Common
     # HTTP Stream used to watch changes on entities
     class WatchStream
-      def initialize(uri, http_options, formatter:)
+      def initialize(uri, options, formatter:)
         @uri = uri
-        @http_client = nil
-        @http_options = http_options
-        @http_options[:http_max_redirects] ||= Kubeclient::Client::DEFAULT_HTTP_MAX_REDIRECTS
+        @options = options
+        @headers = options[:headers]
+        @options[:http_max_redirects] ||= Kubeclient::Client::DEFAULT_HTTP_MAX_REDIRECTS
         @formatter = formatter
+
+        @faraday_client = build_client
       end
 
       def each
         @finished = false
-
-        @http_client = build_client
-        response = @http_client.request(:get, @uri, build_client_options)
-        unless response.code < 300
-          raise Kubeclient::HttpError.new(response.code, response.reason, response)
-        end
-
         buffer = +''
-        response.body.each do |chunk|
-          buffer << chunk
-          while (line = buffer.slice!(/.+\n/))
-            yield(@formatter.call(line.chomp))
+
+        begin
+          @faraday_client.get('', nil, @headers) do |request|
+            request.options.on_data = proc do |chunk|
+              buffer << chunk
+              while (line = buffer.slice!(/.+\n/))
+                yield(@formatter.call(line.chomp))
+              end
+              next if @finished
```

Comment:
The main purpose of `finish` method was to allow interrupting a watch _from another thread_.

The previous implementation by `@http_client&.close` was kinda violent but worked — closing the client would cause the code in `each` to crash in various ways, depending on where exactly it was when it gets closed, and `rescue StandardError` would swallow the crash (when caused by `finish`).

A synchronous `next if @finished` here is much simpler & safer, but I think it'll only stop the loop after it had something to read, which can take unbounded time (until server sends next update)?

Reply:
Hm yes that's a very good observation. I'll try to confirm this with a test and then fix it, hopefully having come up with a solution. ;)